packages/protocol/src/protocol-thp/crypto/curve25519.ts

```diff
@@ -0,0 +1,218 @@
+        throw new Error('Invalid length of coordinate');
+    }
+
+    let array = new Uint8Array(coordinate);
+    array[array.length - 1] &= 0x7f;
+
+    return littleEndianBytesToBigInt(array) % p;
+}
+
+// encodeUCoordinate from
+// https://datatracker.ietf.org/doc/html/rfc7748#section-5
+function encodeCoordinate(coordinate: bigint): Uint8Array {
+    return bigintToLittleEndianBytes(coordinate);
+}
+
+// Returns (second, first) if condition is true and (first, second) otherwise
+// Must be implemented in a way that it is constant time
+function conditionalSwap(a: bigint, b: bigint, condition: boolean): [bigint, bigint] {
+    const mask = condition ? -1n : 0n;
+    const newA = (a & ~mask) | (b & mask);
+    const newB = (a & mask) | (b & ~mask);
+
+    return [newA, newB];
+}
+
+// https://hyperelliptic.org/EFD/g1p/auto-montgom-xz.html#ladder-ladd-1987-m-3
+// (x4, z4) = 2 * (x2, z2)
+// (x5, z5) = (x2, z2) + (x3, z3)
+// where (x1, 1) = (x3, z3) - (x2, z2)
+function ladderOperation(
+    x1: bigint,
+    x2: bigint,
+    z2: bigint,
+    x3: bigint,
+    z3: bigint,
+): [bigint, bigint, bigint, bigint] {
+    const a = (x2 + z2) % p;
+    const aa = (a * a) % p;
+    const b = (x2 - z2 + p) % p;
+    const bb = (b * b) % p;
+    const e = (aa - bb + p) % p;
+    const c = (x3 + z3) % p;
+    const d = (x3 - z3 + p) % p;
+    const da = (d * a) % p;
+    const cb = (c * b) % p;
+    const t0 = (da + cb) % p;
+    const x5 = (t0 * t0) % p;
+    const t1 = (da - cb + p) % p;
+    const t2 = (t1 * t1) % p;
+    const z5 = (x1 * t2) % p;
+    const x4 = (aa * bb) % p;
+    const t3 = (a24 * e) % p;
+    const t4 = (bb + t3) % p;
+    const z4 = (e * t4) % p;
+
+    return [x4, z4, x5, z5];
+}
+
+// X25519 from
+// https://datatracker.ietf.org/doc/html/rfc7748#section-5
+export function curve25519(privateKey: Uint8Array, publicKey: Uint8Array): Buffer {
+    const k = decodeScalar(privateKey);
+    const u = decodeCoordinate(publicKey);
+
+    let x1 = u;
+    let x2 = 1n;
+    let z2 = 0n;
+    let x3 = u;
+    let z3 = 1n;
+    let swap = 0;
+
+    for (let i = 255; i >= 0; i--) {
+        const bit = Number((k >> BigInt(i)) & 1n);
+        swap ^= bit;
+        [x2, x3] = conditionalSwap(x2, x3, Boolean(swap));
+        [z2, z3] = conditionalSwap(z2, z3, Boolean(swap));
+        swap = bit;
+        [x2, z2, x3, z3] = ladderOperation(x1, x2, z2, x3, z3);
+    }
+
+    [x2, x3] = conditionalSwap(x2, x3, Boolean(swap));
+    [z2, z3] = conditionalSwap(z2, z3, Boolean(swap));
+
+    const x = (pow(z2, p - 2n, p) * x2) % p;
+
+    return Buffer.from(encodeCoordinate(x));
+}
+
+// Returns second if condition is true and first otherwise
+// Must be implemented in a way that it is constant time
+function conditionalMove(first: bigint, second: bigint, condition: boolean): bigint {
+    const trueMask = condition ? -1n : 0n;
+    const falseMask = ~trueMask;
+
+    return (first & falseMask) | (second & trueMask);
+}
```

Comment:
I'm not sure if it makes sense to try to make this function run in constant time in TypeScript.

If so, then the line `const trueMask = condition ? -1n : 0n;` is likely not constant time since it probably uses a conditional branch.
if not so, then the function can be defined as `return condition ? second : first;`.